update README

victorbenichoux opened the pull request:
The documentation formatting was a little outdated 😄 , so I suggest a rewrite in markdown format. 

In a first commit I rename the README.txt->README.md, then I split the README (to keep the legacy comments) and rewrite a more informative README for `noahong` objects.

Commit:
readme: rename to .md

Commit:
readme: rewrite documentation

- rewrite NoAho and Mapped documentation
- move all legacy documentation to README-legacy.md and
link to it from the main README.md

Commit:
setup: s/readme.txt/readme.md

pquentin commented on README-legacy.md:
```diff
@@ -1,4 +1,7 @@
+# Legacy README
+
+
+## Differences with `NoAho`
```
Not sure why, but this isn't rendered as a title: https://github.com/victorbenichoux/noahong/blob/readme-format/README-legacy.md

pquentin commented on README-legacy.md:
```diff
@@ -1,4 +1,7 @@
-noahong is a fork https://github.com/JDonner/NoAho
+# Legacy README
```
I don't think we actually want to keep that in the long run?

victorbenichoux replied:
I did not want to cause a loss of information so I kept the original file and linked to it. We may want to amend/delete it later but it felt out of scope

pquentin commented on README.md:
````diff
@@ -0,0 +1,216 @@
+For example, using the above `trie`:
+
+```
+trie.find_short("something foo")
+# returns (10, 13, 'id_foo')
````
What do you think about the way the original README presented examples? I think it's safe to assume the reader as seen a Python REPL

```suggestion
>>> trie.find_short("something foo")
(10, 13, 'id_foo')
```

victorbenichoux replied:
Call me lazy, but I always like to be able to just copy/paste code into my REPL and see what happens, which typically the `>>>` prevent. Although they do in fact look much better.

pquentin commented on README.md:
```diff
@@ -0,0 +1,216 @@
+trie = NoAho()
+trie.add("foo", 0)
+trie.add("bar", CustomClass())
+trie.add("baz", lambda x: x)
```
lambda aren't pickable, so it would be interesting to add a line or two to explain why it works here

victorbenichoux replied:
Added

pquentin commented on README.md:
```diff
@@ -0,0 +1,216 @@
+
+## Mapped `NoAho`
+
+In order to save memory, `noahong` exposes a `Mapped` matching object which can be written to disk and later loaded to perform matches with a smaller memory footprint. 
```
```suggestion
In order to save memory, `noahong` exposes a `Mapped` matching object which can be written to disk and later loaded directly from memory to perform matches with a smaller memory footprint. 
```

victorbenichoux replied:
Fixed

pquentin commented on README.md:
````diff
@@ -0,0 +1,216 @@
+```
+
+
+### Python 2
````
Nope, we don't support Python 2 ! We only support Python 3.6+

It would be interesting to mention that we support macOS, Linux, Windows on Python 3.6+ near the top of the readme

victorbenichoux replied:
Done

pquentin commented on README.md:
````diff
@@ -0,0 +1,216 @@
+
+```
+python3 setup.py install
+```
````
Not sure how useful this is? But if it is, I prefer `pip install .`, and I guess we sould say "can be installed from the sources" or something like that

victorbenichoux replied:
Not sure, it was here before.

Do you think that it is worth mentioning `python cython-regenerate-noaho-setup.py build_ext --inplace` ?

pquentin commented on setup.py:
```diff
@@ -28,7 +29,7 @@
     description="Fast, non-overlapping simultaneous multiple keyword search",
-    long_description=open("README.txt").read(),
+    long_description=open("README.md").read(),
     long_description_content_type="text/plain",
```
```suggestion
    long_description_content_type="text/markdown",
```

pquentin commented on README.md:
```diff
@@ -0,0 +1,216 @@
+
+In order to save memory, `noahong` exposes a `Mapped` matching object which can be written to disk and later loaded to perform matches with a smaller memory footprint. 
+
+The `Mapped` object exposes different finding method and only supports integer payloads.
```
```suggestion
The `Mapped` object exposes different finding methods and only supports integer payloads.
```

victorbenichoux replied:
Added

pquentin commented on README-legacy.md:
```diff
@@ -8,8 +11,7 @@ point, or partially repurpose the library. You might still find it useful.
 
-
-How does it differ from the original project:
+How does it differs from the original project:
```
```suggestion
How does it differ from the original project:
```

victorbenichoux replied:
Fixed

pquentin commented on README-legacy.md:
````diff
@@ -60,64 +59,61 @@ Anti-Features:
+```
+from noaho import NoAho
+trie = NoAho()
+# 'text' below applies to str and unicode in Python 2, or unicode in Python 3 (all there is)
````
Now that we don't support Python 2, I think this line can only add confusion, since `str` is unicode

victorbenichoux replied:
I was unsure about this, should we simply remove all mentions of Python2 then?

pquentin replied:
Yes, please!

victorbenichoux replied:
Done

pquentin commented on README.md:
````diff
@@ -0,0 +1,216 @@
+
+The first thing to do is to instantiate a `NoAho` object and add some keys to it (optionally with different payloads for each).
+
+```
````
All those code blocks could benefit from syntax highligthing

```suggestion
```python3
```

victorbenichoux replied:
Done

pquentin reviewed (changes requested):
Thanks! That's a real improvement.

Commit:
README: add syntax highlighting

Commit:
README-legacy: fix header rendering

victorbenichoux committed:
setup: update readme content type

Co-authored-by: Quentin Pradet <quentin@pradet.me>

Commit:
README: minor improvements

- reflect the drop of Python2 support
- mention supported platforms
- fox typos

Commit:
REAMDE: add note on non serializable payload

victorbenichoux commented:
Thanks for the review Quentin, PTAL!

Commit:
readme: remove python 2 mentions in legacy README

pquentin reviewed (approved):
Thanks!
